### PR TITLE
RawSpeed: Canon EOS 80D: add support for mRaw and sRaw formats (fixes #11028)

### LINKED ISSUE #11028: Cannot compile darktable on Ubuntu 21.10
After installing all required dependencies (`apt build-dep darktable`) and cloning submodules, after executing `./build.sh --asan` I got this build problem
```
In function ‘dt_ioppr_generate_profile_info’,
    inlined from ‘dt_ioppr_add_profile_info_to_list’ at /home/rafal/test/darktable/src/common/iop_profile.c:753:21,
    inlined from ‘dt_ioppr_add_profile_info_to_list’ at /home/rafal/test/darktable/src/common/iop_profile.c:743:1:
/home/rafal/test/darktable/src/common/iop_profile.c:682:11: error: ‘dt_colorspaces_get_matrix_from_output_profile’ accessing 64 bytes in a region of size 16 [-Werror=stringop-overflow=]
  682 |        || dt_colorspaces_get_matrix_from_output_profile(rgb_profile, profile_info->matrix_out,
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  683 |                                                         profile_info->lut_out[0], profile_info->lut_out[1],
      |                                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  684 |                                                         profile_info->lut_out[2], profile_info->lutsize))
      |                                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
```

System specs(command `inxi -GSC -xx`)
```
System:    Host: rafalkom Kernel: 5.13.0-27-generic x86_64 bits: 64 compiler: gcc v: 11.2.0 Desktop: GNOME 40.5 
           tk: GTK 3.24.30 wm: gnome-shell dm: GDM3 Distro: Ubuntu 21.10 (Impish Indri) 
CPU:       Info: Quad Core model: Intel Core i7-4770 bits: 64 type: MT MCP arch: Haswell rev: 3 cache: L2: 8 MiB 
           flags: avx avx2 lm nx pae sse sse2 sse3 sse4_1 sse4_2 ssse3 vmx bogomips: 54276 
           Speed: 2059 MHz min/max: 800/3900 MHz Core speeds (MHz): 1: 2059 2: 1537 3: 3021 4: 3838 5: 1793 6: 3870 7: 2409 
           8: 1789 
Graphics:  Device-1: NVIDIA GM204 [GeForce GTX 970] vendor: Gigabyte driver: nvidia v: 495.46 bus-ID: 01:00.0 
           chip-ID: 10de:13c2 
           Display: x11 server: X.Org 1.20.13 compositor: gnome-shell driver: loaded: nvidia 
           unloaded: fbdev,modesetting,nouveau,vesa resolution: 1: 1440x900~60Hz 2: 1920x1080~60Hz s-dpi: 96 
           OpenGL: renderer: NVIDIA GeForce GTX 970/PCIe/SSE2 v: 4.6.0 NVIDIA 495.46 direct render: Yes 
```

### PR DESCRIPTION
**Pull request by stloeffler**
New decoder hint: wrapped_cr2_slices

### CODE REVIEW

**Comment by LebedevRI** [comment]
Thank you for your contribution!
Can you also check, does the `Canon EOS 7D Mark II` have this same format?
https://redmine.darktable.org/issues/10235

File: src/external/rawspeed/RawSpeed/Cr2Decoder.cpp [hunk 1]
```diff
@@ -179,13 +179,27 @@ RawImage Cr2Decoder::decodeRawInternal() {
...
   bool flipDims = false;
+  bool wrappedCr2Slices = false;
   if (raw->hasEntry((TiffTag)0xc6c5)) {
     ushort16 ss = raw->getEntry((TiffTag)0xc6c5)->getInt();
     // sRaw
     if (ss == 4) {
       mRaw->dim.x /= 3;
       mRaw->setCpp(3);
       mRaw->isCFA = false;
+      // Fix for Canon 80D mraw format.
```
Comment (LebedevRI) [inline src/external/rawspeed/RawSpeed/Cr2Decoder.cpp]: Nitpicking: new line before `// Fix for Canon 80D mraw format.`

File: src/external/rawspeed/RawSpeed/LJpegPlain.cpp [hunk 2]
```diff
@@ -252,8 +252,16 @@ void LJpegPlain::decodeScanLeftGeneric() {
   x = maxSuperH;
   pixInSlice -= maxSuperH;
 
+
   uint32 cw = (frame.w - skipX);
-  for (uint32 y = 0;y < (frame.h - skipY);y += maxSuperV) {
+  uint32 ch = (frame.h - skipY);
+  // Fix for Canon 80D mraw format.
```
Comment (LebedevRI) [inline src/external/rawspeed/RawSpeed/LJpegPlain.cpp]: Nitpicking: newline before `// Fix for Canon 80D mraw format.`

File: src/external/rawspeed/RawSpeed/LJpegPlain.cpp [hunk 3]
```diff
@@ -252,8 +252,16 @@ void LJpegPlain::decodeScanLeftGeneric() {
...
+  // In that format, `frame` is 4032x3402, while `mRaw` is 4536x3024.
+  // Consequently, the slices in `frame` wrap around (this is taken care of by
+  // `offset`) and must be decoded fully (without skipY) to fill the image
+  if (mWrappedCr2Slices)
+    ch = frame.h;
+  for (uint32 y = 0;y < ch;y += maxSuperV) {
```
Comment (LebedevRI) [inline src/external/rawspeed/RawSpeed/LJpegPlain.cpp]: And before the loop

File: src/external/rawspeed/RawSpeed/LJpegPlain.cpp [hunk 4]
```diff
@@ -401,7 +409,14 @@ void LJpegPlain::decodeScanLeft4_2_0() {
   pixInSlice -= 2;
 
   uint32 cw = (frame.w - skipX);
-  for (uint32 y = 0;y < (frame.h - skipY);y += 2) {
+  uint32 ch = (frame.h - skipY);
+  // Fix for Canon 80D mraw format.
+  // In that format, `frame` is 4032x3402, while `mRaw` is 4536x3024.
+  // Consequently, the slices in `frame` wrap around (this is taken care of by
+  // `offset`) and must be decoded fully (without skipY) to fill the image
+  if (mWrappedCr2Slices)
```
Comment (LebedevRI) [inline src/external/rawspeed/RawSpeed/LJpegPlain.cpp]: same here

File: src/external/rawspeed/RawSpeed/LJpegPlain.cpp [hunk 5]
```diff
@@ -625,7 +647,14 @@ void LJpegPlain::decodeScanLeft2Comps() {
   uint32 pixInSlice = slice_width[0] - 1;  // Skip first pixel
 
   uint32 x = 1;                            // Skip first pixels on first line.
-  for (uint32 y = 0;y < (frame.h - skipY);y++) {
+  uint32 ch = (frame.h - skipY);
+  // Fix for Canon 80D mraw format.
+  // In that format, `frame` is 4032x3402, while `mRaw` is 4536x3024.
+  // Consequently, the slices in `frame` wrap around (this is taken care of by
+  // `offset`) and must be decoded fully (without skipY) to fill the image
+  if (mWrappedCr2Slices)
```
Comment (LebedevRI) [inline src/external/rawspeed/RawSpeed/LJpegPlain.cpp]: and here

File: src/external/rawspeed/RawSpeed/LJpegPlain.cpp [hunk 6]
```diff
@@ -721,8 +750,15 @@ void LJpegPlain::decodeScanLeft3Comps() {
 
   uint32 cw = (frame.w - skipX);
   uint32 x = 1;                            // Skip first pixels on first line.
-
-  for (uint32 y = 0;y < (frame.h - skipY);y++) {
+  uint32 ch = (frame.h - skipY);
+  // Fix for Canon 80D mraw format.
```
Comment (LebedevRI) [inline src/external/rawspeed/RawSpeed/LJpegPlain.cpp]: and here

File: src/external/rawspeed/RawSpeed/LJpegPlain.cpp [hunk 7]
```diff
@@ -833,7 +869,14 @@ void LJpegPlain::decodeScanLeft4Comps() {
...
 
-  for (uint32 y = 0;y < (frame.h - skipY);y++) {
+  uint32 ch = (frame.h - skipY);
+  // Fix for Canon 80D mraw format.
+  // In that format, `frame` is 4032x3402, while `mRaw` is 4536x3024.
+  // Consequently, the slices in `frame` wrap around (this is taken care of by
+  // `offset`) and must be decoded fully (without skipY) to fill the image
+  if (mWrappedCr2Slices)
+    ch = frame.h;
+  for (uint32 y = 0;y < ch;y++) {
```
Comment (LebedevRI) [inline src/external/rawspeed/RawSpeed/LJpegPlain.cpp]: and here

**Comment by dabbill** [comment]
@LebedevRI I can confirm that this does fix DT for sRAW and mRAW for the Canon 80D.

**Comment by LebedevRI** [comment]
@stloeffler thank you!

**Pull request merged**